File: 3 Kyu/3 Kyu - Esolang Interpreters #4 - Boolfuck Interpreter.cpp
#include <iostream>
#include <cstdlib>

unsigned loop (const char *code, unsigned index) {
    int pile = 0;
    const bool fwd = code[index] == '[' ? true : false;

    while (code[index]) {
       pile += code[index] == '[' - code[index] == ']';
       if (pile == 0) return index;

       fwd == true ? index++ : index--;
    }
}
char *boolfuck (const char *code, const char *input) {

  char *output = new char [1024], *out = output;
  unsigned index = 0, bit = 0, bin = 0;
  bool tape[30000] = {0}, *data = &tape[30000 / 2] ;

  while (code[index]) {
      switch (code[index]) {
          case '+' : *data ^= 1; break;
          case ',' : *data = input[bin / 8] >> (bin % 8) &1;
                     bin++;
                     break;

          case ';' : output[bit / 8] |= *data << (bit % 8);
                     bit++;
                     break;

          case '<' : data--; break;
          case '>' : data++; break;
          case '[' : if (*data == 0) index = loop (code,index); break;
          case ']' : if (*data != 0) index = loop (code,index); break;
          default : break;
      }
      index++;
    }

    return output;
}

int main () {

  const char *code = ";;;+;+;;+;+;+;+;+;+;;+;;+;;;+;;+;+;;+;;;+;;+;+;;+;+;;;;+;+;;+;;;+;;+;+;+;;;;;;;+;+;;+;;;+;+;;;+;+;;;;+;+;;+;;+;+;;+;;;+;;;+;;+;+;;+;;;+;+;;+;;+;+;+;;;;+;+;;;+;+;+;";

  std::cout << boolfuck(code,"Hello, world!\n");

  int data = 9;

  for (int bit = 0; bit < 8; bit++) {
      int ex = data >> bit &1;
  }


}

File: 3 Kyu/3 Kyu - Esolang Interpreters #4 - Boolfuck Interpreter.rs
use std::collections::VecDeque;

fn find_next (code: &Vec<char>, pos:usize) -> usize {

    let size = code.len();
    let mut index = pos;
    let mut pile = 0;
    let mut fwrd:bool = true;
    if code[index] == ']' { fwrd = false };

    while index < size {
        if code[index] == '[' { pile += 1 }
        if code[index] == ']' { pile -= 1 }

        if pile == 0 { return index }
        if fwrd == true { index +=1 } else { index -= 1 }
    }

    return index;
}
fn boolfuck(code: &str, input: Vec<u8>) -> Vec<u8> {

    let code:Vec<char> = code.chars().collect();
    let mut os:Vec<u8> = Vec::new();
    let mut tape = VecDeque::from(vec![0;1]);
    let mut it = 0;
    let mut index = 0;
    let mut bin = 0;
    let mut out = 0;

    while index < code.len() {

        match code[index] {
            '+' => tape[it] ^= 1,
            '<' => {
                    if it == 0 { tape.push_front(0) } else { it -= 1 }
                },
            '>' => {
                    it += 1;
                    if it == tape.len() { tape.push_back(0) }
                },
            ',' => {
                    tape[it] = input[bin / 8] >> (bin % 8) &1;
                    bin += 1;
                },
            ';' => {
                    if out / 8 >= os.len() { os.push (0) }
                    os[out / 8] |= tape[it] << (out % 8);
                    out += 1
                },
            '[' => if tape[it] == 0 { index = find_next (&code, index) },
            ']' => if tape[it] != 0 { index = find_next (&code, index) },
             _  => (),
        }

        index += 1;
    }

    return os;
}
